### PR TITLE
Bug 1279553 - Fix join and leave chat messages

### PR DESCRIPTION
**Pull request by daveccrit**
WIP @Standard8, @Mardak 
My work so far. I will try and finish this tomorrow when I have some time, unless someone else wants to pick it up and finish it.

### CODE REVIEW

File: shared/js/participantStore.js [hunk 1]
```diff
@@ -119,6 +120,19 @@ loop.store.ParticipantStore = function() {
         }
       }
 
+      if (!updatedParticipant.hasNotifiedOfJoin &&
+          updatedParticipant.isHere &&
+          updatedParticipant.participantName &&
+          Date.now() - updatedParticipant.localPingTime <= PING_TIME) {
```
Comment (Standard8) [inline shared/js/participantStore.js]: There's another instance of this calculation in this file, so we should break this out to some sort of helper function.

File: shared/js/participantStore.js [hunk 2]
```diff
@@ -119,6 +120,19 @@ loop.store.ParticipantStore = function() {
         }
       }
 
+      if (!updatedParticipant.hasNotifiedOfJoin &&
+          updatedParticipant.isHere &&
+          updatedParticipant.participantName &&
```
Comment (Standard8) [inline shared/js/participantStore.js]: Are we always guaranteed a name? Maybe have a fallback just in case someone does something nasty...

Comment (Standard8) [inline shared/js/participantStore.js]: As in, set up with no name & try to sneak into the room. Unlikely, but possible ;-)

File: shared/js/participantStore.js [hunk 3]
```diff
@@ -138,9 +165,18 @@ loop.store.ParticipantStore = function() {
...
+              localPingTime: participants.localPingTime
+            })) {
             participant.isHere = false;
             expiredAny = true;
+            // Chat has already been notified of this participants join. When
+            // participant presence expires, user is notified of participant
+            // has left
+            if (participant.hasNotifiedOfJoin) {
+              this._sendLeaveNotification(participant.participantName, false);
+              participant.hasNotifiedOfJoin = false;
```
Comment (Mardak) [inline shared/js/participantStore.js]: Any reason this flag is set outside of the `_sendLeaveNotification` helper?

File: shared/js/participantStore.js [hunk 4]
```diff
@@ -174,6 +210,30 @@ loop.store.ParticipantStore = function() {
...
 
     /**
+     * Handle Participant Join/Leave Message Notification.
+     *
+     * @param participantName
+     * @param hangup
+     */
+    _sendLeaveNotification(participantName, hangup) {
+      this.dispatcher.dispatch(new sharedActions.RemotePeerLeftChat({
+        participantName: participantName,
```
Comment (Mardak) [inline shared/js/participantStore.js]: Will this end up saying "undefined has left" if there's no name?

File: locale/en-US/shared.properties [hunk 5]
```diff
@@ -38,9 +38,10 @@ rooms_room_join_label=Join the conversation
 rooms_room_joined_owner_connected_label2=Your friend is now connected and will be able to see your tabs.
 rooms_room_joined_owner_not_connected_label=Your friend is waiting to browse {{roomURLHostname}} with you.
 
-peer_left_session=Your friend has left.
-peer_unexpected_quit=Your friend has unexpectedly disconnected.
-peer_join_session=Your friend has joined.
+peer_left_session2={{participantName}} has left.
+peer_unexpected_quit2={{participantName}}  has unexpectedly disconnected.
```
Comment (Standard8) [inline locale/en-US/shared.properties]: nit: double space

File: shared/js/participantStore.js [hunk 6]
```diff
@@ -119,12 +120,38 @@ loop.store.ParticipantStore = function() {
...
 
     /**
+     * Determines if Presence has expired
+     *
+     * @param actionData
+     * @returns {boolean}
+     */
+    _isPresenceExpired(actionData) {
+      if (Date.now() - actionData.localPingTime > PING_TIME) {
+       return true;
```
Comment (Standard8) [inline shared/js/participantStore.js]: Might as well simplify this down:
`return Date.now() - actionData.localPingTime > PING_TIME;`

File: shared/js/dispatcher.js [hunk 7]
```diff
@@ -92,7 +92,7 @@ loop.Dispatcher = (function() {
       this._active = true;
 
       if (this._debug) {
-        console.log("[Dispatcher] Dispatching action", action);
+        console.log("[Dispatcher] Dispatching action", action, action.name);
```
Comment (Standard8) [inline shared/js/dispatcher.js]: The name is already in the action object, so I don't see the need for this? Or are you trying to compensate for the browser console in e10s mode?

If you're compensating, then I'd suggest putting the action.name first before the action object.

File: shared/js/participantStore.js (hunk 3 above)
Comment (Standard8) [inline shared/js/participantStore.js]: We should move these into the function.

File: shared/js/participantStore.js (hunk 6 above)
Comment (Mardak) [inline shared/js/participantStore.js]: Are we actually passing in `actionData`? It looks more like a `participant` but why not just the time?

**Comment by daveccrit** [comment]
Not Completed. There seem to be some issues with events not being dispatched correctly. Will need more work or some future commit may fix this issue.

**Comment by Standard8** [comment]
Support for Hello/Loop has been discontinued.

https://support.mozilla.org/kb/hello-status

Hence closing the old bugs. Thank you for your support.

**Pull request closed without merging**